tests/framework/builder.py

```diff
@@ -40,11 +56,99 @@ def build(self, kernel: Artifact, disks: [DiskArtifact], config: Artifact):
+
+        # Configure logging.
+        response = vm.logger.put(
+            log_path=vm.create_jailed_resource(log_fifo.path),
+            level='Debug',
+            show_level=True,
+            show_log_origin=True,
+        )
+
+        assert vm.api_session.is_status_no_content(response.status_code)
+        # Hardlink all the snapshot files into the microvm jail.
+        jailed_mem = vm.create_jailed_resource(snapshot.mem)
+        jailed_vmstate = vm.create_jailed_resource(snapshot.vmstate)
+        _jailed_disk = vm.create_jailed_resource(snapshot.disks[0])
+        vm.ssh_config['ssh_key_path'] = snapshot.ssh_key
+
+        tap = vm.create_tap_and_ssh_config(host_ip=host_ip,
+                                           guest_ip=guest_ip,
+                                           netmask_len=netmask_len,
+                                           tapname="tap0")
+
+        response = vm.snapshot_load.put(mem_file_path=jailed_mem,
+                                        snapshot_path=jailed_vmstate,
+                                        diff=enable_diff_snapshots)
+
+        assert vm.api_session.is_status_no_content(response.status_code)
+
+        if resume:
+            # Resume microvm
+            response = vm.vm.patch(state='Resumed')
+            assert vm.api_session.is_status_no_content(response.status_code)
+
+        # Return a resumed microvm.
+        return vm, tap
+
+
+class SnapshotBuilder:  # pylint: disable=too-few-public-methods
+    """Create a snapshot from a running microvm."""
+
+    def __init__(self, microvm):
+        """Initialize the snapshot builder."""
+        self._microvm = microvm
+
+    def create(self,
+               disks: [DiskArtifact],
+               ssh_key: Artifact,
+               snapshot_type: SnapshotType = SnapshotType.FULL):
+        """Create a Snapshot object from a microvm and artifacts."""
+        # The APIs for snapshot related procedures does not get timed (yet?).
+        self._microvm.api_session.untime()
+        chroot_path = self._microvm.jailer.chroot_path()
+        snapshot_dir = os.path.join(chroot_path, "snapshot")
+        Path(snapshot_dir).mkdir(parents=True, exist_ok=True)
+        cmd = 'chown {}:{} {}'.format(self._microvm.jailer.uid,
+                                      self._microvm.jailer.gid,
+                                      snapshot_dir)
+        utils.run_cmd(cmd)
+        self._microvm.pause_to_snapshot(
+            mem_file_path="/snapshot/vm.mem",
+            snapshot_path="/snapshot/vm.vmstate",
+            diff=snapshot_type == SnapshotType.DIFF)
+
+        # Create a copy of the ssh_key artifact.
+        ssh_key_copy = ssh_key.copy()
+        mem_path = os.path.join(snapshot_dir, "vm.mem")
+        vmstate_path = os.path.join(snapshot_dir, "vm.vmstate")
+        return Snapshot(mem=mem_path,
+                        vmstate=vmstate_path,
+                        # TODO: support more disks.
```

Comment:
Is this solved by adding all the disks inside the jailer or is something else missing?

Reply:
We need to figure out how we actually store the snapshot artifacts in S3 before finding the right solution here.

Reply:
nit: I would add a bit more detail to the TODO saying that we need to figure this out.